### ISSUE TITLE
Build deterministically

### ISSUE THREAD

**Issue by Hainish** [PyO3/maturin#346]
Tested with
Python 3.6, 3.7, 3.8
Pip 18.1

When building wheels, the generated files do not have [bit-by-bit determinism](https://reproducible-builds.org/docs/deterministic-build-systems/).  Deterministic builds are important for security, they allow third parties to independently verify that the build system has produced a binary faithful to the source code.  In order to make builds deterministic, we would have to determine the sources of entropy in the build process and ensure that these are overwritten by our own deterministic sources.

For instance, we have done this for the zipfile module in python 3.6 here:

https://github.com/EFForg/https-everywhere/blob/master/utils/zipfile_deterministic.py

```bash
(.venv) user@rust ~/workspace/https-everywhere-mitmproxy/rust (master) $ maturin build --manylinux off
🔗 Found pyo3 bindings
🐍 Found CPython 3.6m at python3.6, CPython 3.7m at python3.7, CPython 3.8 at python3.8
    Finished dev [unoptimized + debuginfo] target(s) in 0.23s
📦 Built wheel for CPython 3.6m to /home/user/workspace/https-everywhere-mitmproxy/rust/target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp36-cp36m-linux_x86_64.whl
   Compiling pyo3 v0.11.1
   Compiling https-everywhere-mitmproxy-pyo v0.1.0 (/home/user/workspace/https-everywhere-mitmproxy/rust)
    Finished dev [unoptimized + debuginfo] target(s) in 12.22s
📦 Built wheel for CPython 3.7m to /home/user/workspace/https-everywhere-mitmproxy/rust/target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp37-cp37m-linux_x86_64.whl
   Compiling pyo3 v0.11.1
   Compiling https-everywhere-mitmproxy-pyo v0.1.0 (/home/user/workspace/https-everywhere-mitmproxy/rust)
    Finished dev [unoptimized + debuginfo] target(s) in 13.31s
📦 Built wheel for CPython 3.8 to /home/user/workspace/https-everywhere-mitmproxy/rust/target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp38-cp38-linux_x86_64.whl
(.venv) user@rust ~/workspace/https-everywhere-mitmproxy/rust (master) $ md5sum target/wheels/*
49871e2ad4fa14305e49659a9bee1d90  target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp36-cp36m-linux_x86_64.whl
8cbebf034d3b4c6a25b8990a5f062c6a  target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp37-cp37m-linux_x86_64.whl
1ff96366b6f7fc9300626cab3f187a27  target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp38-cp38-linux_x86_64.whl
(.venv) user@rust ~/workspace/https-everywhere-mitmproxy/rust (master) $ maturin build --manylinux off
🔗 Found pyo3 bindings
🐍 Found CPython 3.6m at python3.6, CPython 3.7m at python3.7, CPython 3.8 at python3.8
   Compiling pyo3 v0.11.1
   Compiling https-everywhere-mitmproxy-pyo v0.1.0 (/home/user/workspace/https-everywhere-mitmproxy/rust)
    Finished dev [unoptimized + debuginfo] target(s) in 14.94s
📦 Built wheel for CPython 3.6m to /home/user/workspace/https-everywhere-mitmproxy/rust/target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp36-cp36m-linux_x86_64.whl
   Compiling pyo3 v0.11.1
   Compiling https-everywhere-mitmproxy-pyo v0.1.0 (/home/user/workspace/https-everywhere-mitmproxy/rust)
    Finished dev [unoptimized + debuginfo] target(s) in 11.08s
📦 Built wheel for CPython 3.7m to /home/user/workspace/https-everywhere-mitmproxy/rust/target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp37-cp37m-linux_x86_64.whl
   Compiling pyo3 v0.11.1
   Compiling https-everywhere-mitmproxy-pyo v0.1.0 (/home/user/workspace/https-everywhere-mitmproxy/rust)
    Finished dev [unoptimized + debuginfo] target(s) in 10.80s
📦 Built wheel for CPython 3.8 to /home/user/workspace/https-everywhere-mitmproxy/rust/target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp38-cp38-linux_x86_64.whl
(.venv) user@rust ~/workspace/https-everywhere-mitmproxy/rust (master) $ md5sum target/wheels/*
5970f3825a7b0dd8e7e39ee5430dc3b6  target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp36-cp36m-linux_x86_64.whl
dde03181e0c1dc98f55ac71cccf814af  target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp37-cp37m-linux_x86_64.whl
8cf322192a8091257a480e606b1b4415  target/wheels/https_everywhere_mitmproxy_pyo-0.1.0-cp38-cp38-linux_x86_64.whl
```

**Comment by dbanty** [comment]
Naive question here: wouldn't maturin building deterministically require rustc to build deterministically? I thought I read recently that it doesn't always do that [but they're working on it](https://github.com/rust-lang/rust/issues/75362)

**Comment by Hainish** [comment]
@dbanty that seems right

**Comment by Hainish** [comment]
Seems solved when building with rust 1.48. Closing
